File: MathCore.DSP/Filters/ButterworthFilter.cs
﻿using System.Runtime.Serialization;

namespace MathCore.DSP.Filters;

/// <summary>Фильтр Баттерворта</summary>
[KnownType(typeof(ButterworthLowPass))]
[KnownType(typeof(ButterworthBandPass))]
[KnownType(typeof(ButterworthBandStop))]
[KnownType(typeof(ButterworthHighPass))]
public abstract class ButterworthFilter : AnalogBasedFilter
{
    /// <summary>Получить список полюсов нормированного фильтра</summary>
    /// <param name="N">Число полюсов</param>
    /// <param name="EpsP">Затухание фильтра</param>
    /// <param name="W0">Множитель коэффициента затухания</param>
    /// <returns>Массив полюсов нормированного фильтра</returns>
    /// <exception cref="ArgumentOutOfRangeException">Если число полюсов меньше 1</exception>
    protected static IEnumerable<Complex> GetNormPoles(int N, double EpsP, double W0 = 1)
    {
        if (N <= 0) throw new ArgumentOutOfRangeException(nameof(N), N, "Число полюсов должно быть больше 0");

        var r = N % 2; // Нечётность порядка фильтра

        // Радиус окружности размещения полюсов фильтра
        var alpha = W0 * EpsP.Pow(-1d / N);

        // Если порядок фильтра нечётный, то первым добавляем центральный полюс
        if (r != 0) yield return -alpha;
        // Расчёт полюсов
        // Угловой шаг между полюсами
        for (var (i, th0) = (r, Consts.pi05 / N); i < N; i += 2)
        {
            var z = Complex.Exp(alpha, th0 * (i - r + N + 1));
            yield return z;
            yield return z.ComplexConjugate;
        }
    }

    protected static IEnumerable<Complex> GetNormPolesGp(int N, double Gp, double W0 = 1)
    {
        if (N <= 0) throw new ArgumentOutOfRangeException(nameof(N), N, "Число полюсов должно быть больше 0");

        var r = N % 2; // Нечётность порядка фильтра

        // Радиус окружности размещения полюсов фильтра
        var alpha = W0 * (1 / (Gp * Gp) - 1).Pow(-0.5 / N);

        // Если порядок фильтра нечётный, то первым добавляем центральный полюс
        if (r != 0) yield return -alpha;
        // Расчёт полюсов
        // Угловой шаг между полюсами
        for (var (i, th0) = (r, Consts.pi05 / N); i < N; i += 2)
        {
            var z = Complex.Exp(alpha, th0 * (i - r + N + 1));
            yield return z;
            yield return z.ComplexConjugate;
        }
    }

    /// <summary>Инициализация фильтра Баттерворта</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    protected ButterworthFilter(double[] B, double[] A, Specification Spec) : base(B, A, Spec) { }
}

File: MathCore.DSP/Filters/Python/filters/butterworth_filter.py
# Базовый класс фильтра Баттерворта (порт C# ButterworthFilter)
import numpy as np
from .analog_based_filter import AnalogBasedFilter, Specification

_pi05 = np.pi / 2  # π/2


class ButterworthFilter(AnalogBasedFilter):
    """Базовый класс фильтра Баттерворта с вычислением нормированных полюсов."""

    def __init__(self, B: np.ndarray, A: np.ndarray, spec: Specification):
        super().__init__(B, A, spec)

    @staticmethod
    def get_norm_poles(N: int, EpsP: float, W0: float = 1.0) -> list:
        """
        Нормированные полюса фильтра Баттерворта.

        Полюса размещаются равномерно на окружности радиуса alpha в левой
        полуплоскости p-плоскости.

        Параметры
        ---------
        N    : порядок фильтра
        EpsP : неравномерность АЧХ в полосе пропускания (EpsP = sqrt(1/Gp² - 1))
        W0   : масштабирующий множитель (нормирующая частота)
        """
        if N <= 0:
            raise ValueError(f"Число полюсов должно быть больше 0, получено N={N}")

        r = N % 2           # нечётность порядка
        alpha = W0 * EpsP ** (-1.0 / N)  # радиус окружности полюсов
        th0 = _pi05 / N     # угловой шаг между полюсами

        poles = []
        if r != 0:
            poles.append(complex(-alpha, 0.0))  # центральный полюс нечётного порядка
        for i in range(r, N, 2):
            angle = th0 * (i - r + N + 1)
            z = alpha * np.exp(1j * angle)
            poles.append(z)
            poles.append(z.conjugate())
        return poles

    @staticmethod
    def get_norm_poles_gp(N: int, Gp: float, W0: float = 1.0) -> list:
        """
        Нормированные полюса Баттерворта с явным указанием Gp.

        Параметры
        ---------
        N  : порядок фильтра
        Gp : коэффициент передачи на границе пропускания (0 < Gp ≤ 1)
        W0 : масштабирующий множитель
        """
        if N <= 0:
            raise ValueError(f"Число полюсов должно быть больше 0, получено N={N}")

        r = N % 2
        # alpha = W0 * (1/Gp² - 1)^{-0.5/N}
        alpha = W0 * (1.0 / (Gp * Gp) - 1.0) ** (-0.5 / N)
        th0 = _pi05 / N

        poles = []
        if r != 0:
            poles.append(complex(-alpha, 0.0))
        for i in range(r, N, 2):
            angle = th0 * (i - r + N + 1)
            z = alpha * np.exp(1j * angle)
            poles.append(z)
            poles.append(z.conjugate())
        return poles
